Sources/SwiftDocC/Infrastructure/Link Resolution/PathHierarchy.swift

```diff
@@ -122,14 +122,18 @@ struct PathHierarchy {
             
             var topLevelCandidates = nodes
             for relationship in graph.relationships where relationship.kind.formsHierarchy {
-                guard let sourceNode = nodes[relationship.source] else {
+                guard let sourceNode = nodes[relationship.source], let expectedContainerName = sourceNode.symbol?.pathComponents.dropLast().last else {
                     continue
                 }
-                if let targetNode = nodes[relationship.target] {
+                // The relationship only specify the target symbol's USR but if the target symbol has different representations in different source languages the relationship
+                // alone doesn't specify which language representation the source symbol belongs to. We could check the source and target symbol's interface language but that
+                // would require that we redundantly create multiple nodes for the same symbol in many common cases and then merge them. To avoid doing that, we instead check
+                // the source symbol's path components to find the correct target symbol by matching its name.
+                if let targetNode = nodes[relationship.target], targetNode.name == expectedContainerName {
                     targetNode.add(symbolChild: sourceNode)
                     topLevelCandidates.removeValue(forKey: relationship.source)
                 } else if let targetNodes = allNodes[relationship.target] {
-                    for targetNode in targetNodes {
+                    for targetNode in targetNodes where targetNode.name == expectedContainerName {
```

Comment:
What would happen if expectedContainerName from above is nil? (Can a symbol graph relationship source node not have a symbol?)

Also, is it possible for none of the target nodes to have a name that matches the source symbol's path component?

Maybe these possible corner conditions are what caused the crash/assertion in the first place, from PR https://github.com/apple/swift-docc/pull/759 ?

Reply:
> What would happen if expectedContainerName from above is nil? (Can a symbol graph relationship source node not have a symbol?)

If is expected that the source symbol for a relationship is in the same symbol graph file as that relationship. If it's not then we ignore that relationship. This happens in [this guard statement above](https://github.com/apple/swift-docc/pull/765/files/5d5cadc60951098a045fa90d74a76d8b2b963845#diff-e0d0f88acc1bca134d47867a977e60a988a20118c4b52451f90376f6e3d56f7cR125).

It is also expected that a symbol's path components describe the path to that symbol (excluding the module name). Unfortunately path components alone can be very ambiguous when building up a hierarchy of symbols so DocC prioritizes symbol relationships to build up the hierarchy. 

There is a fallback mechanism [more towards the end of this initializer](https://github.com/apple/swift-docc/pull/765/files/5d5cadc60951098a045fa90d74a76d8b2b963845#diff-e0d0f88acc1bca134d47867a977e60a988a20118c4b52451f90376f6e3d56f7cR183-R209) that processes every symbol that hasn't been added to the hierarchy yet and tries to find a suitable location based only on its path components. If that can't find an existing symbol with the right name it will create an unfindable "sparse node" with that name and add the symbol and its sparse symbol to the hierarchy.

> Also, is it possible for none of the target nodes to have a name that matches the source symbol's path component?

Yes, see above.

> Maybe these possible corner conditions are what caused the crash/assertion in the first place, from PR #759 ?

No. Both the crash in #759 and this assertion are caused by the opposite problem; the symbol was added to multiple target nodes. 

This issue wouldn't normally happen. In a simple single language setup there would only be one symbol with each USR so there would only be one possible target node to add it to. In a normal multi language setup there could be different language representations a symbol but DocC would always prefer to add the symbol as a member of the target symbol from the same symbol graph file, so the source symbol would only be added to one target node.

What made both #759 and this assertion happen is that documentation was built in a non-standard setup where the main module was built together with the extensions symbol graph files of other modules that were extending the main module. DocC's intended support for extensions builds the symbol graphs for the main module together with the extension symbol graph files for the other modules that the main module extends, not together with the extension symbol graph files from the other modules that extends the main module. 

In this non-standard setup, the source symbol isn't defined in the main module but it is added to a symbol in the main module. Because of this it's not clear which language representation of the target symbol it should be added to. Previously DocC added it to both which sometimes led to the crash in #759 (depending on which target symbol it was added to last). Now DocC only adds it to one of the target symbols.

Reply:
Thanks for the detailed explanation!